### PR TITLE
Migrate to Bit.ly API v4

### PR DESCRIPTION
**Pull request by isuruhettiarachchi**
Resolves wso2/product-is#14084

TODO: Update test cases

### CODE REVIEW

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java [hunk 1]
```diff
@@ -263,4 +273,43 @@ protected Map<ClaimMapping, String> getSubjectAttributes(OAuthClientResponse tok
         }
         return claims;
     }
+
+    @Override
+    protected String sendRequest(String url, String accessToken) throws IOException {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Claim URL: " + url);
```
Comment (omindu) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:281]: What does it mean by the claim URL ?

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java [hunk 2]
```diff
@@ -263,4 +273,43 @@ protected Map<ClaimMapping, String> getSubjectAttributes(OAuthClientResponse tok
         }
         return claims;
     }
+
+    @Override
+    protected String sendRequest(String url, String accessToken) throws IOException {
```
Comment (omindu) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:278]: Does this method is used to invoke anything other than the user info endpoint ?

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java (hunk 1 above)
Comment (omindu) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:281]: If so, can we consider adding the authenticator identifier to the debug logs ?

ex: `log.debug("Bitly user info response: " + builder.toString());`

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java (hunk 2 above)
Comment (omindu) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:278]: If so, can we consider adding the authenticator identifier to the debug logs ?

ex: `log.debug("Bitly user info response: " + builder.toString());`

Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:278]: This method is used in bitly authenticator only to invoke user info endpoint only. We can modify the debug log

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java (hunk 1 above)
Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:281]: We can modify this debug log to show `log.debug("URL: " + url)`.

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java (hunk 2 above)
Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:278]: Fixed

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java (hunk 1 above)
Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java]: Fixed

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java [hunk 3]
```diff
@@ -263,4 +273,43 @@ protected Map<ClaimMapping, String> getSubjectAttributes(OAuthClientResponse tok
...
+
+        if (log.isDebugEnabled()) {
+            log.debug("URL: " + url);
```
Comment (nilasini) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:281]: Shall we mention "URL"  as user info url of bitly?

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java [hunk 4]
```diff
@@ -263,4 +273,43 @@ protected Map<ClaimMapping, String> getSubjectAttributes(OAuthClientResponse tok
...
+        }
+
+        if (url == null) {
```
Comment (nilasini) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:284]: Use stringUtlils.isBlank

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java [hunk 5]
```diff
@@ -263,4 +273,43 @@ protected Map<ClaimMapping, String> getSubjectAttributes(OAuthClientResponse tok
...
+                inputLine = reader.readLine();
+            }
+        } finally {
+            if (reader != null) {
+                reader.close();
+            }
+        }
+
+        if (log.isDebugEnabled() && IdentityUtil.isTokenLoggable(IdentityConstants.IdentityTokens.USER_ID_TOKEN)) {
+            log.debug("Bitly user info response: " + builder.toString());
```
Comment (nilasini) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:311]: We are not printing the token then why are we checking isTokenLoggable?

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java (hunk 3 above)
Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java]: Fixed

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java (hunk 5 above)
Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java]: Yea, we can remove this condition

Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java]: Fixed

File: component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java [hunk 6]
```diff
@@ -263,4 +271,43 @@ protected Map<ClaimMapping, String> getSubjectAttributes(OAuthClientResponse tok
...
+                inputLine = reader.readLine();
+            }
+        } finally {
+            if (reader != null) {
+                reader.close();
+            }
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("Bitly user info response: " + builder);
```
Comment (nilasini) [inline component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:309]: Check claims loggable and add the log

Comment (bhagyasakalanka) [reply component/src/main/java/org/wso2/carbon/identity/authenticator/bitly/BitlyAuthenticator.java:310]: Fixed

**Review by nilasini — APPROVED** [review]

**Comment by nilasini** [comment]
Make sure to add the test case with another PR @bhagyasakalanka

**Pull request merged**